rasa/core/lock.py

```diff
@@ -0,0 +1,160 @@
+import json
+import logging
+import typing
+from collections import deque
+from typing import Text, Optional, Union, Deque, Dict, Any
+
+import time
+
+if typing.TYPE_CHECKING:
+    pass
+
+logger = logging.getLogger(__name__)
+
+
+class Ticket:
+    def __init__(self, number: int, expires: float):
+        self.number = number
+        self.expires = expires
+
+    def has_expired(self):
+        return time.time() > self.expires
+
+    def dumps(self) -> Text:
+        """Return json dump of `Ticket`."""
+
+        return json.dumps(dict(number=self.number, expires=self.expires))
+
+    @classmethod
+    def from_dict(cls, data: Dict[Text, Any]) -> "Ticket":
+        """Creates `Ticket` from dictionary."""
+
+        return cls(data.get("number"), data.get("expires"))
+
+    def __repr__(self):
+        return "Ticket(number: {}, expires: {})".format(self.number, self.expires)
+
+
+class TicketLock(object):
+    def __init__(
+        self, conversation_id: Text, tickets: Optional[Deque[Ticket]] = None
+    ) -> None:
+        self.conversation_id = conversation_id
+        self.tickets = tickets or deque()
+
+    @classmethod
+    def from_dict(cls, data: Dict[Text, Any]) -> "TicketLock":
+        """Create `TicketLock` from dictionary."""
+
+        tickets = [Ticket.from_dict(json.loads(d)) for d in data.get("tickets")]
+        return cls(data.get("conversation_id"), deque(tickets))
+
+    def dumps(self) -> Text:
+        """Return json dump of `TicketLock`."""
+
+        tickets = [ticket.dumps() for ticket in self.tickets]
+        return json.dumps(dict(conversation_id=self.conversation_id, tickets=tickets))
+
+    def is_locked(self, ticket_number: int) -> bool:
+        """Return whether `ticket_number` is locked.
+
+        Returns False if lock has expired.
+        Otherwise returns True if `now_serving` is not equal to `ticket`.
+        """
+
+        return self.now_serving != ticket_number
+
+    def issue_ticket(self, lifetime: Union[float, int]) -> int:
+        """Issue a new ticket and return its number."""
+
+        self.remove_expired_tickets()
+        number = self.last_issued + 1
+        ticket = Ticket(number, time.time() + lifetime)
+        self.tickets.append(ticket)
+
+        return number
+
+    def remove_expired_tickets(self) -> None:
+        """Remove expired tickets."""
+
+        # iterate over copy of self.tickets so we can remove items
+        for ticket in list(self.tickets):
+            if ticket.has_expired():
+                self.tickets.remove(ticket)
+
+    @property
+    def last_issued(self) -> int:
+        """Return number of the ticket that was last added.
+
+        Return -1 if no tickets exist.
+        """
+
+        ticket_number = self._ticket_number_for_index(-1)
+        if ticket_number is not None:
+            return ticket_number
+
+        return -1
```

Comment:
should we at least define a constant for that: E.g. `NO_TICKET_ISSUED` or sth like that?

Reply:
not sure about this, as it's the only place where it's used

Reply:
but would still make the code more readable, wouldn't it? can also just name it in the function if you want, but having a name for it would help.

Reply:
I find this more confusing to be honest 😛 
![image](https://user-images.githubusercontent.com/20581300/61891873-ed765600-af0a-11e9-9002-59a58d4ada58.png)

Reply:
😞 Then let's skip it. Docstring should be fine.